pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,199 @@
+		delete(ipc.ipToIdentityCache, endpointIP)
+		// avoid deletion from nil map
+		if _, found2 := ipc.identityToIPCache[identity]; found2 {
+			delete(ipc.identityToIPCache[identity], endpointIP)
+			if len(ipc.identityToIPCache[identity]) == 0 {
+				delete(ipc.identityToIPCache, identity)
+			}
+		}
+		ipc.xdsResourceMutator.Delete(envoy.NetworkPolicyHostsTypeURL, identity.StringID(), false)
+	}
+}
+
+// LookupByIP returns the corresponding security identity that endpoint IP maps
+// to within the provided IPCache, as well as if the corresponding entry exists
+// in the IPCache.
+func (ipc *IPCache) LookupByIP(endpointIP string) (identity.NumericIdentity, bool) {
+	ipc.Mutex.RLock()
+	identity, exists := ipc.ipToIdentityCache[endpointIP]
+	ipc.Mutex.RUnlock()
+	return identity, exists
+}
+
+// IPIdentityMappingOwner is the interface the owner of an identity allocator
+// must implement
+type IPIdentityMappingOwner interface {
+	// TriggerPolicyUpdates will be called whenever a policy recalculation
+	// must be triggered
+	TriggerPolicyUpdates(force bool) *sync.WaitGroup
+}
+
+// GetIPIdentityMapModel returns all known endpoint IP to security identity mappings
+// stored in the key-value store.
+func GetIPIdentityMapModel() {
+	// TODO (ianvernon) return model of ip to identity mapping. For use in CLI.
+}
+
+func ipIdentityWatcher(owner IPIdentityMappingOwner) {
+	watcher := kvstore.ListAndWatch("endpointIPWatcher", IPIdentitiesPath, 512)
+	for {
+		var (
+			identity     identity.NumericIdentity
+			cacheChanged bool
+		)
+
+		// Get events from channel as they come in.
+		event := <-watcher.Events
+
+		_ = json.Unmarshal(event.Value, &identity)
+
+		// Synchronize local caching of endpoint IP to identity mapping with
+		// operation key-value store has informed us about.
+
+		cachedIdentity, exists := IPIdentityCache.LookupByIP(event.Key)
+
+		switch event.Typ {
+		case kvstore.EventTypeCreate, kvstore.EventTypeModify:
+			// Update local cache
+			if !exists || cachedIdentity != identity {
+				IPIdentityCache.Upsert(event.Key, identity)
+				cacheChanged = true
+			}
+		case kvstore.EventTypeDelete:
+			if exists {
+				IPIdentityCache.Delete(event.Key)
+				cacheChanged = true
+			}
+		}
+
+		// Trigger policy updates only if cache has changed.
+		if cacheChanged {
+			log.WithFields(logrus.Fields{
+				"endpoint-ip":      event.Key,
+				"cached-identity":  cachedIdentity,
+				logfields.Identity: identity,
+			}).Debugf("triggering policy updates because endpoint IP cache changed state")
+			owner.TriggerPolicyUpdates(true)
+		}
+	}
+}
+
+// InitIPIdentityWatcher initializes the watcher for ip-identity mapping events
+// in the key-value store.
+func InitIPIdentityWatcher(owner IPIdentityMappingOwner) {
+	setupIPIdentityWatcher.Do(func() {
+		go ipIdentityWatcher(owner)
+	})
+}
+
+// SetAddressSpace sets the value of this package's AddressSpace to addressSpace.
+func SetAddressSpace(addressSpace string) {
+	AddressSpace = addressSpace
```

Comment:
`AddressSpace` can be private, no?

Reply:
It's accessed in the controllers for the endpoint `pkg/endpoint/policy.go` so I made it public.